docs/proposals/petset.md

```diff
@@ -0,0 +1,533 @@
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# PetSets: Running pods which need strong identity and storage
+
+## Open Issues
+
+* Add examples
+* Discuss failure modes for various types of clusters
+* Provide an active-active example
+* Templating proposals need to be argued through to reduce options
+
+## Motivation
+
+Many examples of clustered software systems require stronger guarantees per instance than are provided
+by the Replication Controller (aka ReplicaSet). Instances of these systems typically require:
+
+1. Data per instance which should not be lost even if the pod is deleted, typically on a persistent volume
+   * Some cluster instances may have tens of TB of stored data - forcing new instances to replicate data
+     from other members over the network is onerous
+2. A stable and unique identity associated with that instance of the storage - such as a unique member id
+3. A consistent network identity that allows other members to locate the instance even if the pod is deleted
+4. A predictable number of instances to ensure that systems can form a quorum
+   * This may be necessary during initialization
+5. Instances can migrate from machine to machine as necessary and are not tied to an instance
+6. The ability to scale up in a controlled fashion, but are very rarely scaled down without human
+   intervention
+
+Kubernetes should expose a pod controller (a PetSet) that satisfies these requirements in a flexible
+manner. It should be easy for users to manage and reason about the behavior of this set. An administrator
+with familiarity in a particular cluster system should be able to leverage this controller and its
+supporting documentation to run that clustered system on Kubernetes. It is expected that some adaptation
+is required to support each new cluster.
+
+
+## Use Cases
+
+The software listed below forms the primary use-cases for a PetSet on the cluster - problems encountered
+while adapting these for Kubernetes should be addressed in a final design.
+
+* Quorum with Leader Election
+  * MongoDB - in replica set mode forms a quorum with an elected leader, but instances must be preconfigured
+    and have stable network identities.
+  * ZooKeeper - forms a quorum with an elected leader, but is sensitive to cluster membership changes and
+    replacement instances *must* present consistent identities
+  * etcd - forms a quorum with an elected leader, can alter cluster membership in a consistent way, and
+    requires stable network identities
+* Decentralized Quorum
+  * Cassandra - allows flexible consistency and distributes data via innate hash ring sharding, is also
+    flexible to scaling, more likely to support members that come and go. Scale down may trigger massive
+    rebalances.
+* Active-active
+  * **Need active-active master example - galera? **
```

Comment:
Elasticsearch is active-active (unless there's some meaning of that I don't grasp). It replicates shards between all masters for redundancy. One is elected leader by a quorum, but in terms of storage it is just like any other instance.

Elasticsearch presents a number of pain points I'm not sure are covered here yet. 

- When an instance goes down and comes back, standard procedure is to wipe any index shards that have been updated and resync from another node.
- Thus upgrading a cluster should really include cutting off writes to the cluster, bringing down all the instances, doing the upgrade, and then starting them again (and allowing writes). Otherwise instances will thrash the network with furious re-syncing.
- If changing the cluster size, you also need to change related parameters like the quorum for starting and the quorum for electing a leader. 
 - These parameters can be changed via http request, which would require some kind of per-instance hot-deploy hook to make the request;
 - or you can take down the entire cluster to change the parameters via env vars or whatever.
- Elastic recommends using local SSDs as storage for performance reasons. For that to work, the instances have to be carefully placed on intended hosts - allowed sole access to a host volume that persists. If you do an instance or cluster restart, there is no identity tying the instance to a node, but the same set of nodes should be used. And only one instance can be allowed to mount a volume at a time, otherwise they'll stomp each other's storage.